### PR TITLE
Cut in typeclasses eauto: hint extern when implementation

### PR DESCRIPTION
**Pull request by mattam82**
This provides a new "Hint Extern self? priority pattern? (when iftac)? => thentac" hint kind allowing to implement full backtracking and cuts in (typeclasses) eauto resolution, giving access to the tclIFCATCH primitive of the proof-engine.

Two new things:
- Using self allows the hint extern to completely drive the proof search at this point: the tactic may choose to call self to continue resolution on the remaining subgoals but also to use it in a nested way to solve goals it produces itself. If one puts self but no `when` clause, then it can just be used to do recursive calls of proof search respecting the depth/hint databases etc... of the ongoing proof search. If one doesn't call `self` on resulting subgoals of an extern application, then it is run on the resulting subgoals, as before.
- If `when iftac` is provided then the hint acts as `tclIFCATCH iftac thentac`. This is a little bit more restrictive than a `tclCASE` which could also decide to do something specific ratner than continue resolution if the `iftac` has no success, but still allows to implement (hard) cut: if `iftac` succeeds at least once, `thentac` is executed and no backtracking on the "brother" subgoals of resolution can happen (in other words the failure continuation is dropped). One can use `once` and `self` in iftac and thentac to implement additional soft cuts.

[ ] documentation
[ ] changelog entry

### CODE REVIEW

File: tactics/auto.ml
Comment (Zimmi48) [inline tactics/auto.ml]: Can't you translate the comment rather than fixing French grammar? ^^

Comment (Zimmi48) [reply tactics/auto.ml]: OK maybe not now given that there is a whole paragraph to translate.

**Comment by robbertkrebbers** [comment]
What's the motivation for this feature? I have to say I cannot make much sense of the test case as part of this PR.

**Comment by Zimmi48** [comment]
I think this follows this: https://github.com/coq/ceps/pull/30/commits/0bf88ec7f8238f6435f2f60864357573e664d883?short_path=1622ffd#diff-1622ffdb2217392a0159ba8e8b3e5563

**Comment by maximedenes** [comment]
@mattam82 what is your schedule on this one?

**Comment by maximedenes** [comment]
@mattam82 ping

**Comment by gares** [comment]
In order to review and to document this PR it would help to give a rephrasing in Coq's Type Classes language of the following Prolog code:
```prolog
b(1).
b(2).
c(2,1).
e(1).
a(X,Y) :- b(X), c(X,Y), !, e(X). % fail
a(X,Y) :- d(X,Y).  % cut away

?- a(A,B). % should fail
```
The trace should be

Goal | rule
------|-----
a(A,B) |  a(X,Y) :- b(X)....
b(A) | b(1)
c(1,B) | fail
b(A) | b(2)
c(2,B) | c(2,1)
! | ...
e(2) | fail

**Review by gares — DISMISSED** [review]
To my understanding:
1. the new feature lets you code a prolog cut
2. the user API is very low level
3.  ~~it is non compositional~~
4. the code is not very clean

Description:
1. is all positive
2. I think that exposing the LogicT IFTE brick is too low level. First of all, given our public is not the one of logic programming, "if" can only be called this way when combined with ONCE in the evaluation of the guard, while the API that is given to the user is "error prone" in the sense that one may forget the once (or simply don't know how logic programming works or not having read the ICFP paper). I guess this if fixable by providing (also/only) a higher level API with a true IF, and rename/remove this one.
3. ~~is unfixable AFAICT, it is a shortcoming of LogicT. Unless I missed something, the "cut" is not going to affect instances that are declared separately to (later than) the one(s) declared within the hint extern if. This is also the tricky part of the test I asked to write. I think this problem should be documented, if possible even detected at instance declaration time (if there is a hint-if on that class, and you give a precedence that puts the new instance after the hint-if, then the cut is not going to cut it)~~ @CohenCyril explained me that everything is in CPS, including the program itself, so the cut can indeed cut what is left.
4. the code is reusing Extern for both regular Extern and IF. Then 2 out of 3 users of the data type ignore the Extern-IF. I'm not very knowledgeable of the implementation of Hints, but I suspect one could make a new node for this type of hint, a node to TCeauto handles (the other users of Hints can then complain loudly)

**Comment by gares** [comment]
I'd like to request a review (from the user perspective) by @CohenCyril , esp look at the file(s) added to the test suite

**Comment by gares** [comment]
> What's the motivation for this feature? I have to say I cannot make much sense of the test case as part of this PR.

Dear @robbertkrebbers if you look at the Prolog code I wrote above you see a `!` that means "when this point is reached, cut away all proof tree branches that could lead here". It is a semantically horrible construct of logic programming that is necessary in order to express a smart search strategy. Sometimes you know that "any other solution would be equivalent", so there is no point in trying all the others. Or you know that there are other solutions, maybe different, but you don't care since the one you just found is good enough for you. With my programmer hat on and the pun stick in my hand, I'd say that a Prolog program not using `!` is just exponential.

This is for @ppedrot : Elpi has a pretty printer: clauses not containing ! are printed in red ;-)

**Comment by robbertkrebbers** [comment]
> With my programmer hat on and the pun stick in my hand, I'd say that a Prolog program not using ! is just exponential.

Yes, we figured that out the hard way. See also #6714, and in particular the comment

https://github.com/coq/coq/issues/6714#issuecomment-364418761

that provides a workaround to have something like the `!`.

**Comment by CohenCyril** [comment]
> I'd like to request a review (from the user perspective) by @CohenCyril , esp look at the file(s) added to the test suite

I'll take a look

**Comment by CohenCyril** [comment]
> Yes, we figured that out the hard way. See also #6714, and in particular the comment
> 
> #6714 (comment)
> 
> that provides a workaround to have something like the !.

This is also related to the discussion that happened here: https://github.com/coq/ceps/pull/30

**Comment by maximedenes** [comment]
@CohenCyril did you have a look?

**Comment by maximedenes** [comment]
@mattam82 do you still plan to have this ready by tomorrow? if not, please update the milestone.

**Comment by CohenCyril** [comment]
> @CohenCyril did you have a look?

not yet

**Comment by CohenCyril** [comment]
@maximedenes  I'll have a look at it this month. Ping me in april if it is still not done.

**Comment by ppedrot** [comment]
@CohenCyril While we're still barely in April, I'll take the occasion to ping you back in lieu of @maximedenes.

**Comment by CohenCyril** [comment]
Hi @ppedrot, thanks for pinging me, I am scheduling it for this week!

**Review by ppedrot — COMMENTED** [review]
For some reason I have to review this, but I'm not specialist enough on the whatever-auto tactics to make any useful comment regarding the semantics. I'll leave that to others.

File: tactics/auto.ml
Comment (mattam82) [reply tactics/auto.ml]: Indeed the treatment of the then is deferred to the special treatment of Extern in class_tactics.ml:L475. It is not the case for other hints, I can add a warning when introducing such hints telling they can only be used with tceauto and ignore them in `auto/eauto`, does that sound ok?

**Comment by CohenCyril** [comment]
So I encoded Enrico's problem, and the trace is ok.
* One problem I have is that I do not know how to do a once on two subgoals but not the third (cf comment down there). Is it easy to do? If not, it should be dealt with before merge.
* Another is that the trace contains many anomalies of the form: `(a ?x) failed with: Anomaly "Uncaught exception Refiner.FailError(0, _)."` which I do not understand. I think this is important to solve, but I do not understand the meaning.

I agree with point (2.) of Enrico that the encoding is too low level to be usable by end-users as a Prolog cut, but now one can define a new vernac to generate the right `Hint Extern If` (with a special syntax for instances with cuts in them, or just a dummy `cut := unit` symbol which can be used as a marker).

I do not understand (3.) as I think I got the behavior you expected: only the precedence of a rule counts, so that rules with lower precedence are cut even if they are defined before... I think this what you expect of a typeclass cut and I would say it is compositional. Did I miss something?

As for (4) since `Hint Extern [If]` is supposed to be a primitive anyway, with a higher interface to insert it in the HindDb later, if `If` is documented in the manual I fine with it.

```
Module cut2.
(* Prolog program *)
(* b(1). *)
(* b(2). *)
(* c(2,1). *)
(* e(1). *)
(* d(3,3). *)
(* a(X,Y) :- b(X), c(X,Y), !, e(X). % fail *)
(* a(X,Y) :- d(X,Y).  % cut away *)
(* ?- a(A,B). % should fail *)

(* Coq translation *)

Class a (x y : nat).
Class b (x : nat).
Class c (x y : nat).
Class d (x y : nat).
Class e (x : nat).

Definition b1 : b 1 := Build_b _.
Hint Extern 0 (b _) => exact b1 : typeclass_instances.
Definition b2 : b 2 := Build_b _.
Hint Extern 1 (b _) => exact b2 : typeclass_instances.
Definition c21 : c 2 1 := Build_c _ _.
Hint Extern 0 (c _ _) => exact c21 : typeclass_instances.
Definition e1 : e 1 := Build_e _.
Hint Extern 0 (e _) => exact e1 : typeclass_instances.
Definition bcea x y `{b x} `{c x y} `{e x} : a x y := Build_a _ _.
Hint Extern 0 (a _ _) If
  notypeclasses refine (@bcea _ _ _ _ _);
  [typeclasses eauto|typeclasses eauto|]
  (* how to do once on those two goals simultaneously? *)
  =>
  typeclasses eauto : typeclass_instances.
Definition da x y `{d x y} : a x y :=  Build_a _ _.
Hint Extern 1 (a _ _) => refine (@da _ _ _) : typeclass_instances.
Definition d33 : d 3 3 := Build_d _ _.
Hint Extern 0 (d _ _) => exact d33 : typeclass_instances.

Goal exists x, exists y, a x y.
Proof.
eexists; eexists.
Fail typeclasses eauto. (* Why so many Anomalies ? *)
Abort.
End cut2.

```

**Comment by CohenCyril** [comment]
Also, if i replace `Hint Extern` (normal ones not `If`s) , the trace does not look good and I do not understand why...

**Comment by gares** [comment]
I think you want `only 1,2: tc eauto`.

**Comment by vbgl** [comment]
Is someone working on this? Will it be ready for 8.10?

**Comment by mattam82** [comment]
I changed the milestone, I won't have time to look at it before the freeze. Anyone is welcome to take it up though.

**Comment by Zimmi48** [comment]
Dear PR author(s) and contributors, it seems to us this PR may need additional work beyond the 8.12 branching deadline on May 15th.

We are thus removing the 8.12+beta1 milestone; please retarget as appropriate, and also consider updating labels to reflect current status, moving the PR to draft status, etc.

We apologize in advance if we misunderstood the PR status.

**Comment by SkySkimmer** [comment]
@mattam82 do you have plans to revive this PR?

**Comment by CohenCyril** [comment]
@mattam82 should we have a meeting on this? And if so who else wants in? (@gares and his prolog expertise?)
(The meeting on namespaces was quite productive, I hope this one could)

**Comment by mattam82** [comment]
I'm not planning to work on it in the immediate future, but sure we can have a chat about it. Let's coordinate on Zulip?

**Comment by maximedenes** [comment]
I'd like to be part of it, since I believe it could address the need behind https://github.com/coq/coq/issues/6714.

**Comment by SkySkimmer** [comment]
I rebased it.

**Comment by mattam82** [comment]
Hint Extern (a _ _) sefl => notypeclasses refine (bcea _ _ _ _ _); once (only 1-2: self); (self + raise Commit)

**Comment by gares** [comment]
```coq
Hint Extern (a _ _) self g =>
  notypeclasses refine (bcea _ _ _ _ _);
  once (only 1-2: self);
  last_call g self

```

**Comment by mattam82** [comment]
After a long discussion about the design of Hint Extern If tac Then ... and possibly more expressive variants: We came to settle on keeping the Hint Extern If which allows to express one "global" cut on brothers in the proof search tree + a reference to "self"/continue for solving subgoals with the right options/depth/db etc. One can then use `once` for internal cuts if needed.

**Comment by mattam82** [comment]
Look at the top post for the currently implemented solution :)

**Comment by Zimmi48** [comment]
Shouldn't this PR be in draft mode for the time being?

**Comment by mattam82** [comment]
@Zimmi48 right

File: dev/ci/ci-hott.sh
Comment (JasonGross) [inline dev/ci/ci-hott.sh:8]:
```suggestion
( cd "${CI_BUILD_DIR}/hott" && coq_makefile -f _CoqProject -o Makefile \
```
But I think also this may fail unless the native compiler is disabled?

File: dev/ci/user-overlays/13952-mattam82-hints-modes.sh
Comment (JasonGross) [inline dev/ci/user-overlays/13952-mattam82-hints-modes.sh:2]: Should not be needed anymore since https://github.com/mit-plv/rewriter/pull/24 was merged

File: dev/ci/ci-hott.sh
Comment (mattam82) [reply dev/ci/ci-hott.sh:8]: Ah I didn't know about that.

File: dev/ci/user-overlays/13952-mattam82-hints-modes.sh
Comment (mattam82) [reply dev/ci/user-overlays/13952-mattam82-hints-modes.sh:2]: Yep, I'll remove that in 13952 if not done already.

File: dev/ci/ci-hott.sh
Comment (mattam82) [reply dev/ci/ci-hott.sh]: Let's try deactivating it too

Comment (JasonGross) [reply dev/ci/ci-hott.sh]: See https://github.com/HoTT/HoTT/issues/1454.  Idk if this is an issue with the native complication scheme, or if it's an issue with having so many things `Defined`

**Comment by gares** [comment]
This is important

**Comment by Zimmi48** [comment]
@gares Reopening is not enough. What you can do is reopen then take care of the rebasing. If you cannot do this right now but really want to give more time for a rebase, you can reopen and remove the stale label. @coqbot will re-add it the next day, but with it will restart the 30 days counter.

**Comment by gares** [comment]
Thanks for the info!

**Comment by Zimmi48** [comment]
What about rebasing it this time?

**Pull request closed without merging**
